dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/CodeGenerator.java

```diff
@@ -285,7 +291,116 @@ public static boolean isValidSHA256HexFormat( String s )
             return false;
         }
 
-        // SHA-256 hexadecimal strings are exactly 64 characters long
+        // SHA-256 in hexadecimal are exactly 64 characters long
         return s.length() == 64;
     }
+
+    private static char generateRandomCharacter( String charSet )
+    {
+        SecureRandom sr = SecureRandomHolder.GENERATOR;
+        return charSet.charAt( sr.nextInt( charSet.length() ) );
+    }
+
+    /**
+     * Generates a random password with the given size, has to be minimum 4
+     * characters long.
+     * <p>
+     * The password will contain at least one digit, one special character, one
+     * uppercase letter and one lowercase letter.
+     * <p>
+     * The password will not contain more than 2 consecutive letters to avoid
+     * generating words.
+     *
+     * @param passwordSize the size of the password.
+     * @return a random password.
+     */
+    public static char[] generateValidPassword( int passwordSize )
```

Comment:
I am wondering if this is just used in tests or also part of production. Since the first 4/5 characters are off a fix type there is little variation for short passwords. Which would be fine in tests but if we generate short ones in some role in production this might be to small of a space.

Reply:
Minimum default length for passwords is 8, and this method it's only used for invites. It also uses a lot of special chars (after 4th char), so entropy is pretty high then. So someone would need to lower the default password length to 4 and do a lot of invites for this to be an issue.